src/ServerHandler.cpp

```diff
@@ -1,8 +1,153 @@
+}
+
+void ServerHandler::parseMessageFrame(Frame messageFrame) {
+    std::string genre = messageFrame.getHeaderVal("destination");
+    std::vector <std::string> subStrs;
+    std::string body = messageFrame.getBody();
+    split(body, subStrs, " ");
+    if (subStrs[1] == "wish") {
+        std::string book = extractBookName(subStrs, 3, subStrs.size()); // starting from the word 'borrow'
+        if (_inventory.isInInventory(genre, book)) {
+            Frame ansFrame = Frame(SEND);
+            ansFrame.addHeader("destination", genre);
+            ansFrame.setBody(_inventory.getUsername() + " has " + book);
+            _connectionHandler->sendFrameAscii(ansFrame.toString(), STOMP_DELIMITER.c_str()[0]);
+        }
+    }
+    else if (subStrs[0] == "Returning" && subStrs[subStrs.size() - 1] == _inventory.getUsername()) {
+        // starting after 'Returning', ending before 'to'
+        std::string book = extractBookName(subStrs, 1, subStrs.size() - 2);
+        _inventory.addBook(genre, book);
+    }
+    else if (subStrs[0] == "Taking" && subStrs[subStrs.size() - 1] == _inventory.getUsername()) {
+        // starting after 'Taking', ending before 'from'
+        std::string book = extractBookName(subStrs, 1, subStrs.size() - 2);
+        _inventory.removeFromInventory(genre, book);
+    }
+    else if (body == "book status") {
+        Frame ansFrame = Frame(SEND);
+        ansFrame.addHeader("destination", genre);
+        std::string body = booksString(_inventory.getGenreBooks(genre));
+        ansFrame.setBody(body);
+        _connectionHandler->sendFrameAscii(ansFrame.toString(), STOMP_DELIMITER.c_str()[0]);
+    }
+    else if (subStrs[1] == "has") {
+        std::string book = extractBookName(subStrs, 2, subStrs.size()); //staring after the word 'has'
+        if (_inventory.isInWishList(genre, book)) {
+            std::string lender = subStrs[0];
+            Frame ansFrame = Frame(SEND);
+            ansFrame.addHeader("destination", genre);
+            ansFrame.setBody("Taking " + book + " from " + lender);
+            _connectionHandler->sendFrameAscii(ansFrame.toString(), STOMP_DELIMITER.c_str()[0]);
+            _inventory.borrowBook(book, lender);
+            _inventory.removeFromWishList(genre, book);
+        }
+    }
+}
+
+void ServerHandler::parseServerFrame() {
+    std::string frameString;
+    _connectionHandler->getFrameAscii(frameString, STOMP_DELIMITER.c_str()[0]);
+    Frame frame = Frame(frameString);
+    FrameCommand cmd = frame.getCommand();
+    switch (cmd) {
+        case MESSAGE: {
+            parseMessageFrame(frame);
+        }
+        default: { // CONNECTED, RECEIPT, and ERROR frames, should be passed to user.
+            sendFrameToClient(frame);
+        }
+    }
+}
+
+void ServerHandler::run() {
+    Frame frameFromClient;
+    while (true) {
+        while ((frameFromClient = receiveFrameFromClient()).getCommand() != UNINITIALIZED) {
+            parseUserFrame(frameFromClient);
+        }
```

Comment:
This isn't good, here the serverHandler waits for a command before even trying to handle a message from the server, but in actuality the serverHandler will handle a lot more frames than the user sends commands

Reply:
that's what we agreed. give priority to user request, + I assume the serverHandler is much faster than user typing, so it will finish dealing with all Server's messages before getting any frame from the client.
you could argue that creating a frame every time is inefficient... do you have an other idea?

Reply:
This isn't priority, the inner while will loop over and over until the user types a command, then deal with one server message and go into that loop again. This is like having one thread.
Also this thread doesn't have any exit condition, When the user disconnects the entire client program should end

Reply:
notice the it checks the frame is different from UNINITIALIZED. so no waiting to message from user.
as for the endless loop. you are correct. I will complete it when i'll write the main loop. notice that client doesn't die with logout, but with "bye"

> Once the client receives the RECEIPT frame, it should close the socket and await further user commands

> You can either close the client on logout/error, or at some "bye" command of your choice. We may check your client with valgrind, close things in an orderly fashion.

Reply:
You are absolutely right, my bad!